c10/util/sparse_bitset.h

```diff
@@ -0,0 +1,857 @@
+ /// fast for the most common cases, SparseBitVector is implemented as a linked
+ /// list of SparseBitVectorElements.  We maintain a pointer to the last
+ /// SparseBitVectorElement accessed (in the form of a list iterator), in order
+ /// to make multiple in-order test/set constant time after the first one is
+ /// executed.  Note that using vectors to store SparseBitVectorElement's does
+ /// not work out very well because it causes insertion in the middle to take
+ /// enormous amounts of time with a large amount of bits.  Other structures that
+ /// have better worst cases for insertion in the middle (various balanced trees,
+ /// etc) do not perform as well in practice as a linked list with this iterator
+ /// kept up to date.  They are also significantly more memory intensive.
+
+ template <unsigned ElementSize = 128> struct SparseBitVectorElement {
+ public:
+   using BitWord = unsigned long;
+   using size_type = unsigned;
+   enum {
+     BITWORD_SIZE = sizeof(BitWord) * CHAR_BIT,
+     BITWORDS_PER_ELEMENT = (ElementSize + BITWORD_SIZE - 1) / BITWORD_SIZE,
+     BITS_PER_ELEMENT = ElementSize
+   };
+
+ private:
+   // Index of Element in terms of where first bit starts.
+   unsigned ElementIndex;
+   BitWord Bits[BITWORDS_PER_ELEMENT];
+
+   SparseBitVectorElement() {
+     ElementIndex = ~0U;
+     memset(&Bits[0], 0, sizeof (BitWord) * BITWORDS_PER_ELEMENT);
+   }
+
+ public:
+   explicit SparseBitVectorElement(unsigned Idx) {
+     ElementIndex = Idx;
+     memset(&Bits[0], 0, sizeof (BitWord) * BITWORDS_PER_ELEMENT);
+   }
+
+   // Comparison.
+   bool operator==(const SparseBitVectorElement &RHS) const {
+     if (ElementIndex != RHS.ElementIndex)
+       return false;
+     for (unsigned i = 0; i < BITWORDS_PER_ELEMENT; ++i)
+       if (Bits[i] != RHS.Bits[i])
+         return false;
+     return true;
+   }
+
+   bool operator!=(const SparseBitVectorElement &RHS) const {
+     return !(*this == RHS);
+   }
+
+   // Return the bits that make up word Idx in our element.
+   BitWord word(unsigned Idx) const {
+     assert(Idx < BITWORDS_PER_ELEMENT);
+     return Bits[Idx];
+   }
+
+   unsigned index() const {
+     return ElementIndex;
+   }
+
+   bool empty() const {
+     for (unsigned i = 0; i < BITWORDS_PER_ELEMENT; ++i)
+       if (Bits[i])
+         return false;
+     return true;
+   }
+
+   void set(unsigned Idx) {
+     Bits[Idx / BITWORD_SIZE] |= 1L << (Idx % BITWORD_SIZE);
+   }
+
+   bool test_and_set(unsigned Idx) {
+     bool old = test(Idx);
+     if (!old) {
+       set(Idx);
+       return true;
+     }
+     return false;
+   }
+
+   void reset(unsigned Idx) {
+     Bits[Idx / BITWORD_SIZE] &= ~(1L << (Idx % BITWORD_SIZE));
+   }
+
+   bool test(unsigned Idx) const {
+     return Bits[Idx / BITWORD_SIZE] & (1L << (Idx % BITWORD_SIZE));
+   }
+
+   size_type count() const {
+     unsigned NumBits = 0;
+     for (unsigned i = 0; i < BITWORDS_PER_ELEMENT; ++i)
+       NumBits += __builtin_popcountl(Bits[i]);
```

Comment:
This won't work on MSVC.  You probably need to pull in the relevant helper function from LLVM to abstract between Clang/GCC/MSVC.

Reply:
Ditto for the other __builtin calls